### PR TITLE
8310176: JDK 11 G1 crash during full GC with +UseStringDeduplication

### PR DESCRIPTION
**Pull request by caoman**
Hi all,

Could anyone review this bug fix to G1's full GC for JDK 11. See https://bugs.openjdk.org/browse/JDK-8310176 for details.
This issue only affects JDK 11 and 12, and is not a backport.

-Man

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310176](https://bugs.openjdk.org/browse/JDK-8310176): JDK 11 G1 crash during full GC with +UseStringDeduplication (**Bug** - P3)

### Reviewers
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**) ⚠️ Review applies to [797f8761](https://git.openjdk.org/jdk11u-dev/pull/1948/files/797f8761dabf83ed74af853c6cd3bc72d5f0a0ba)
 * [Richard Reingruber](https://openjdk.org/census#rrich) (@reinrich - **Reviewer**) ⚠️ Review applies to [797f8761](https://git.openjdk.org/jdk11u-dev/pull/1948/files/797f8761dabf83ed74af853c6cd3bc72d5f0a0ba)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev.git pull/1948/head:pull/1948` \
`$ git checkout pull/1948`

Update a local copy of the PR: \
`$ git checkout pull/1948` \
`$ git pull https://git.openjdk.org/jdk11u-dev.git pull/1948/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1948`

View PR using the GUI difftool: \
`$ git pr show -t 1948`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/1948.diff">https://git.openjdk.org/jdk11u-dev/pull/1948.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk11u-dev/pull/1948#issuecomment-1593967386)

### CODE REVIEW

**Comment by GoeLin** [comment]
Please only label if the PR is ready. 
Please elaborate on you testing. 
Would it be possible to add a jtreg test?

**Comment by caoman** [comment]
Thanks for the quick review. Initially I validated the fix with an internal application that exposes this bug, as I didn't notice any jtreg test failing. With a closer look I discovered the jtreg test `TestStringDeduplicationFullGC.java`, and the flag `-XX:+StringDeduplicationRehashALot`. Running both together can expose this bug. Added a new test that is similar to `TestStringDeduplicationTableResize.java`. 

The rehashing mechanism together with `-XX:+StringDeduplicationRehashALot` were deleted in https://bugs.openjdk.org/browse/JDK-8254598, so this test is also JDK11-only.

**Review by phohensee — APPROVED** [review]
Lgtm.

**Comment by caoman** [comment]
Thank you Paul for the review.
I see the test TestStringDeduplicationTableRehashFullGC.java is failing in pre-submit tests. Interestingly I cannot reproduce the failure locally on Linux x64. I'll try fixing it before requesting with `jdk11u-fix-request` label.

**Comment by caoman** [comment]
Yay, the test finally passed. I'm requesting for approval with `jdk11u-fix-reques` now.

**Review by phohensee — APPROVED** [review]

File: test/hotspot/jtreg/gc/g1/TestStringDeduplicationTools.java [hunk 1]
```diff
@@ -383,6 +383,22 @@ public static void testTableRehash() throws Exception {
...
+    public static void testTableRehashFullGC() throws Exception {
+        // Test with StringDeduplicationRehashALot using full GCs
+        // Table resizing prevents table rehashing from happening, thus we use
+        // SmallNumberOfStrings to limit the number of table resizes.
+        OutputAnalyzer output = DeduplicationTest.run(SmallNumberOfStrings,
+                                                      DefaultAgeThreshold,
+                                                      FullGC,
+                                                      "-Xlog:gc,gc+stringdedup=trace",
+                                                      "-XX:+StringDeduplicationRehashALot");
+        output.shouldContain("Concurrent String Deduplication");
```
Comment (reinrich) [inline test/hotspot/jtreg/gc/g1/TestStringDeduplicationTools.java:395]: Is there a reason you don't check `output.shouldContain("Full GC")`?

**Review by reinrich — APPROVED** [review]
Looks good to me. I've also executed the test in a loop without issues.
Richard.

File: test/hotspot/jtreg/gc/g1/TestStringDeduplicationTools.java (hunk 1 above)
Comment (caoman) [reply test/hotspot/jtreg/gc/g1/TestStringDeduplicationTools.java:395]: Thanks, added this check. I copied it from `testTableRehash()` initially.

**Comment by caoman** [comment]
Thanks for the review! I'll wait for JDK-8310176 to be approved before integrating this.

**Comment by caoman** [comment]
Could anyone help sponsor this change?

**Comment by jianglizhou** [comment]
> Could anyone help sponsor this change?

Man, what specific help are you looking for? If it's for integration, I can help with that.

**Comment by caoman** [comment]
Thanks Jiangli for sponsoring!

**Pull request closed without merging**